File: Fore-C_v2/codes/Run_model_scenarios.R
# Run scenarios using quantile regression models

source("./codes/custom_functions/fun_quant_forest_predict.R")

# open final model objects
source("./codes/Final_covariates_by_disease_and_region.R")

# set up directory filepaths
scenario_dir <- "../compiled_data/scenarios_inputs/"
x <- list.files(scenario_dir, full.names = TRUE)

scenarios_save_dir <- "../compiled_data/scenarios_outputs/"
if(dir.exists(scenarios_save_dir) == FALSE){
  dir.create(scenarios_save_dir)
}

# open input files
lapply(x, load, .GlobalEnv)

# run scenarios --------
# GA GBR
ga_gbr_scenarios <- qf_predict_scenarios(df = ga_gbr_scenarios
                                         , regionGBRtrue = TRUE
                                         , family = "all"
                                         , final_mod = GA_GBR_Model
                                         )

# pre-calculate disease risk change
# may want to use UprEstimate
ga_gbr_scenarios$disease_risk_change <- round((ga_gbr_scenarios$estimate - ga_gbr_scenarios$value) * 10)
# can't decrease more than 100%
ga_gbr_scenarios$disease_risk_change[ga_gbr_scenarios$disease_risk_change < -100] <- -100
# save data to run and then replace with same name
save(ga_gbr_scenarios, file = "../uh-noaa-shiny-app/forec_shiny_app_data/Scenarios/ga_gbr_scenarios.RData")

# WS GBR -------------------------------------
ws_gbr_scenarios <- qf_predict_scenarios(df = ws_gbr_scenarios
                                         , regionGBRtrue = TRUE
                                         , family = "plating"
                                         , final_mod = WS_GBR_Model
                                         )

# pre-calculate disease risk change
ws_gbr_scenarios$disease_risk_change <- round((ws_gbr_scenarios$estimate - ws_gbr_scenarios$value) * 10)
# can't decrease more than 100%
ws_gbr_scenarios$disease_risk_change[ws_gbr_scenarios$disease_risk_change < -100] <- -100
# save data to run and then replace with same name
save(ws_gbr_scenarios, file = "../uh-noaa-shiny-app/forec_shiny_app_data/Scenarios/ws_gbr_scenarios.RData")

# GA Pacific ---------------------------------
ga_pac_scenarios <- qf_predict_scenarios(df = ga_pac_scenarios
                                         , regionGBRtrue = FALSE
                                         , family = "Poritidae"
                                         , final_mod = GA_Pacific_Model
                                         )

# adjust development levels to match shiny app (scaled values)
ga_pac_development_levels_scaled <- seq(from = 0.1, to = 1, by = 0.1)
ga_pac_development_levels <- seq(from = 1, to = 255, length.out = length(ga_pac_development_levels_scaled))# 

for(i in 1:length(ga_pac_development_levels)){
  ga_pac_scenarios$Response_level[
    ga_pac_scenarios$Response == "Development" & 
      ga_pac_scenarios$Response_level == ga_pac_development_levels[i]
    ] <- ga_pac_development_levels_scaled[i]
}

# pre-calculate disease risk change
# may want to use UprEstimate
ga_pac_scenarios$disease_risk_change <- round((ga_pac_scenarios$estimate * 100) - ga_pac_scenarios$value)
# can't decrease more than 100%
ga_pac_scenarios$disease_risk_change[ga_pac_scenarios$disease_risk_change < -100] <- -100
# save data to run and then replace with same name
save(ga_pac_scenarios, file = "../uh-noaa-shiny-app/forec_shiny_app_data/Scenarios/ga_pac_scenarios.RData")

# WS Pacific ---------------------------------
ws_pac_scenarios <- qf_predict_scenarios(df = ws_pac_scenarios
                                         , regionGBRtrue = FALSE
                                         , family = "Acroporidae"
                                         , final_mod = WS_Pacific_Model
                                         )

# pre-calculate disease risk change
# may want to use UprEstimate
ws_pac_scenarios$disease_risk_change <- round((ws_pac_scenarios$estimate * 100) - ws_pac_scenarios$value)
# can't decrease more than 100%
ws_pac_scenarios$disease_risk_change[ws_pac_scenarios$disease_risk_change < -100] <- -100
# save data to run and then replace with same name
save(ws_pac_scenarios, file = "../uh-noaa-shiny-app/forec_shiny_app_data/Scenarios/ws_pac_scenarios.RData")

File: Fore-C_v2/codes/Run_model_scenarios.py
# -*- coding: utf-8 -*-
"""
Created on Thu May 26 16:27:52 2022

@author: jamie
"""

import pandas as pd
import os

# load model objects
from codes.Final_covariates_by_disease_and_region import GA_GBR_Model, GA_Pacific_Model, WS_GBR_Model, WS_Pacific_Model
from codes.custom_functions.fun_quant_forest_predict import qf_predict_scenarios

# load scenarios
ga_gbr_scenarios = pd.read_csv('../compiled_data/scenarios_inputs/ga_gbr_scenarios.csv')
ga_pac_scenarios = pd.read_csv('../compiled_data/scenarios_inputs/ga_pac_scenarios.csv')
ws_gbr_scenarios = pd.read_csv('../compiled_data/scenarios_inputs/ws_gbr_scenarios.csv')
ws_pac_scenarios = pd.read_csv('../compiled_data/scenarios_inputs/ws_pac_scenarios.csv')
 
# output directory path
# scenarios_save_dir = '../compiled_data/scenarios_outputs/'
scenarios_save_dir = '../uh-noaa-shiny-app/forec_shiny_app_data/Scenarios/'

# Check whether the specified path exists or not
isExist = os.path.exists(scenarios_save_dir)

# Create a new directory if it does not exist 
if not isExist:
  os.makedirs(scenarios_save_dir)

# list scenario id variables
scenarios_id_vars = ['ID', 'Latitude', 'Longitude', 'Region', 'predicted', 'Response', 'Response_level']

# run scenarios ---------------------------------------------------------------

# GA GBR ---------------
ga_gbr_scenario_predictions = qf_predict_scenarios(
    df = ga_gbr_scenarios
    , regionGBRtrue = True
    , family = 'all'
    , final_mod = GA_GBR_Model
    , id_vars = scenarios_id_vars
    )

ga_gbr_filepath = scenarios_save_dir + 'ga_gbr_scenarios.csv'
ga_gbr_scenario_predictions.to_csv(ga_gbr_filepath, index = False)


# GA Pacific -----------
ga_pac_scenario_predictions = qf_predict_scenarios(
    df = ga_pac_scenarios
    , regionGBRtrue = False
    , family = 'Poritidae'
    , final_mod = GA_Pacific_Model
    , id_vars = scenarios_id_vars
    )

ga_pac_filepath = scenarios_save_dir + 'ga_pac_scenarios.csv'
ga_pac_scenario_predictions.to_csv(ga_pac_filepath, index = False)

# WS GBR ---------------
ws_gbr_scenario_predictions = qf_predict_scenarios(
    df = ws_gbr_scenarios
    , regionGBRtrue = True
    , family = 'plating'
    , final_mod = WS_GBR_Model
    , id_vars = scenarios_id_vars
    )

ws_gbr_filepath = scenarios_save_dir + 'ws_gbr_scenarios.csv'
ws_gbr_scenario_predictions.to_csv(ws_gbr_filepath, index = False)


# WS Pacific -----------
ws_pac_scenario_predictions = qf_predict_scenarios(
    df = ws_pac_scenarios
    , regionGBRtrue = False
    , family = 'Acroporidae'
    , final_mod = WS_Pacific_Model
    , id_vars = scenarios_id_vars
    )

ws_pac_filepath = scenarios_save_dir + 'ws_pac_scenarios.csv'
ws_pac_scenario_predictions.to_csv(ws_pac_filepath, index = False)

